### PR TITLE
8281560: Matcher.hitEnd returns unexpected results in presence of CANON_EQ flag.

### PR DESCRIPTION
**Pull request by igraves**
Fixing a bug in `NFCCharProperty` where code falling through an if-statement can prematurely set the matcher's `hitEnd` field to true.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8281560](https://bugs.openjdk.java.net/browse/JDK-8281560): Matcher.hitEnd returns unexpected results in presence of CANON_EQ flag.

### Reviewers
 * [Roger Riggs](https://openjdk.java.net/census#rriggs) (@RogerRiggs - **Reviewer**) ⚠️ Review applies to cf8ffd6c308ff097c3cac02dd91b0fb004cc7618
 * [Lance Andersen](https://openjdk.java.net/census#lancea) (@LanceAndersen - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/7466/head:pull/7466` \
`$ git checkout pull/7466`

Update a local copy of the PR: \
`$ git checkout pull/7466` \
`$ git pull https://git.openjdk.java.net/jdk pull/7466/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 7466`

View PR using the GUI difftool: \
`$ git pr show -t 7466`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/7466.diff">https://git.openjdk.java.net/jdk/pull/7466.diff</a>

</details>

### CODE REVIEW

File: src/java.base/share/classes/java/util/regex/Pattern.java [hunk 1]
```diff
@@ -4006,7 +4006,9 @@ boolean match(Matcher matcher, int i, CharSequence seq) {
                 if (j < matcher.to)
                     return false;
             }
-            matcher.hitEnd = true;
+            else {
```
Comment (LanceAndersen) [inline src/java.base/share/classes/java/util/regex/Pattern.java:4009]: Is there a reason the "else" starts on its own line?

**Review by RogerRiggs — APPROVED** [review]
LGTM

File: test/jdk/java/util/regex/RegExTest.java [hunk 2]
```diff
@@ -4557,12 +4557,36 @@ public static void badIntersectionSyntax() {
...
 
+    //This test is for 8281560
+    @Test
+    public static void prematureHitEndInNFCCharProperty() {
+        var testInput = "a1a1";
+        var pat1 = "(a+|1+)";
+        var pat2 = "([a]+|[1]+)";
+
+        var matcher1 = Pattern.compile(pat1).matcher(testInput);
+        var matcher2 = Pattern.compile(pat2).matcher(testInput);
```
Comment (jaikiran) [inline test/jdk/java/util/regex/RegExTest.java:4568]: Hello Ian,
The JBS issue notes that the bug is only noticed when the `Pattern.CANON_EQ` is used. The javadoc of this `CANON_EQ` states that this isn't enabled by default. Do you think this test should also include matchers which use Pattern(s) with this `CANON_EQ` enabled?

File: src/java.base/share/classes/java/util/regex/Pattern.java (hunk 1 above)
Comment (igraves) [reply src/java.base/share/classes/java/util/regex/Pattern.java]: I must love Enter too much. Thanks for the catch!

File: test/jdk/java/util/regex/RegExTest.java (hunk 2 above)
Comment (igraves) [reply test/jdk/java/util/regex/RegExTest.java]: Oh wow yes. Thanks for this catch. I was rewriting the tests to fit this mold and left the flags out. Added them back in. Thanks again!

Comment (jaikiran) [reply test/jdk/java/util/regex/RegExTest.java]: Thank you Ian, the changed version of the test looks good.

**Review by LanceAndersen — APPROVED** [review]

**Pull request closed without merging**